staging/src/k8s.io/component-base/config/serializer/encode.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serializer
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+)
+
+// EncodingFormat is a high-level, typed constant configuring how an object should be encoded.
+type EncodingFormat string
+
+const (
+	// ContentTypeYAML encodes a config file in the YAML format.
+	ContentTypeYAML EncodingFormat = "application/yaml"
+	// ContentTypeJSON encodes a config file in compact JSON format.
+	ContentTypeJSON EncodingFormat = "application/json"
+)
+
+// NewConfigSerializer creates a new implementation of the ConfigSerializer interface.
+func NewConfigSerializer(scheme *runtime.Scheme, codecs *serializer.CodecFactory) ConfigSerializer {
```

Comment:
The strict serializers I can get behind, but I don't view "config" as a natural type of api object that deserves its own serializer. This appears to be a strict serializer that works with either yaml or json, right? Please name it accordingly. Its use isn't going to be limited to config. IOW, name it by what it *is*, not what you're going to use it for.

Reply:
Are you suggesting something like `NewStrictYAMLJSONSerializer`?

Would you want this abstraction to be housed somewhere more generic than `component-base` such as `apimachinery`?

Reply:
That's a mouthful, but yeah, it's an improvement.

I would like to leave it defined here. We can always promote it to a more general package if it grows more users. :)

Reply:
So we'll keep the `component-base/config` package and just rename the symbols.

@stewart-yu feel free to cherry-pick https://github.com/stealthybox/kubernetes/commit/11067af110bc2e48b2c3a5b22be25ab743e398e4 if it looks good

Reply:
done